src/java.base/share/classes/com/sun/crypto/provider/PKCS12PBECipherCore.java

```diff
@@ -351,16 +298,23 @@ void implInit(int opmode, Key key, AlgorithmParameterSpec params,
             Arrays.fill(derivedKey, (byte)0);
 
             try {
-                if (cipherImpl != null && cipherImpl instanceof ARCFOURCipher) {
-                    ((ARCFOURCipher) cipherImpl).engineInit(opmode, cipherKey, random);
-
+                if (cipher instanceof ARCFOURCipher rc4) {
+                    rc4.engineInit(opmode, cipherKey, random);
                 } else {
                     byte[] derivedIv = derive(passwdChars, salt, iCount, 8,
                             CIPHER_IV);
-                    IvParameterSpec ivSpec = new IvParameterSpec(derivedIv, 0, 8);
+                    IvParameterSpec ivSpec = new IvParameterSpec(derivedIv, 0,
+                            8);
 
                     // initialize the underlying cipher
-                    cipher.init(opmode, cipherKey, ivSpec, random);
+                    if (cipher instanceof RC2Cipher rc2) {
+                        rc2.engineInit(opmode, cipherKey, ivSpec, random);
+                    } else if (cipher instanceof DESedeCipher tripleDes) {
+                        tripleDes.engineInit(opmode, cipherKey, ivSpec, random);
+                    } else {
```

Comment:
Currently, the specified CipherSpi object is one of RC4, RC2, DESede. The "else" part is for catching new PKCS12 PBE algorithms support which uses other cipher algorithms.
CipherSpi.engineInit(...) is protected, so that's why we use the instanceof to cast the CipherSpi object into the actual impl class in order to call the engineInit(...) since the actual impl class is in the same package of this class.

Reply:
The `core.init(..., cipher)` is actually `cipher.init(core.translateKeyAndParams())`. Is it possible we write it this way?

Reply:
It's possible, more refactoring would be needed and not necessarily less lines of code. With your suggested change, the caller has to explicitly destroy the derived key after the cipher.engineInit() call. This would be repeated in all PKCS12 PBE cipher impl classes, but then there'd be no casting of the actual classes. I assume this is what you are referring to? Can code it out and see how it looks.

Reply:
If the returned key-and-iv class implements Closeable, then you can do a try-with-resources to destroy the key, which saves you more lines.

Reply:
Yes, good point, will do so.